Refactor for OSGI compliance

mike-mcgann opened the pull request:
Daffodil shares package namespaces across different libraries such that library A can have a class named foo.bar.One and library B can have a class named foo.bar.Two. This can cause issues with class loaders that prohibit different jars from owning the same package (i.e., in an OSGI container). This factoring ensures that each package is owned by a single jar.
    
[DAFFODIL-2683](https://issues.apache.org/jira/browse/DAFFODIL-2683)

stevedlawrence commented on xsrc/test/scripts/refactor/fix-imports.scala:
```diff
@@ -0,0 +1,184 @@
+    "daffodil-io" -> "io",
+    "daffodil-lib" -> "lib",
+    "daffodil-macro-lib" -> "lib",
+    "daffodil-runtime1" -> "runtime1",
+    "daffodil-runtime1-layers" -> "layers.runtime1",
+    "daffodil-runtime1-unparser" -> "unparsers.runtime1"
```
Maybe runtime1 wants to be `org.apache.daffodil.runtime1.core` or `org.apache.daffodil.runtime1.parsers`? That way all the runtime1 stuff are in subpackages of the same parent package.

mbeckerle replied:
Core is the short name for the primary "schema compiler" module currently. I think using core for the compiler aspects of other modules is fine.

Maybe everything "core" should be renamed "comp" instead though.

mike-mcgann replied:
Both `daffodil-runtime1` and `daffodil-runtime1-layers` contain an `org.apache.daffodil.layers` package. Maybe `daffodil-runtime1` stays as-is and `daffodil-runtime1-layers` becomes `org.apache.daffodil.layers.impl`?

mbeckerle replied:
The implementation of layers is in daffodil-runtime1
The predefined layers that come built-in to Daffodil are in daffodil-runtime1-layers. 

So I would think daffodil-runtime1's layers should be the one that becomes layers.impl.

mike-mcgann replied:
Selecting a good name is always a challenging task. I'm neutral on the matter. Out of these three (or with another name not yet listed), let me know which one we should use and I'll update the scripts .

stevedlawrence commented on xsrc/test/scripts/refactor/fix-imports.scala:
```diff
@@ -0,0 +1,184 @@
+    "daffodil-core" -> "core",
+    "daffodil-io" -> "io",
+    "daffodil-lib" -> "lib",
+    "daffodil-macro-lib" -> "lib",
```
This might want to be `org.apache.daffodil.macros`? This package isn't actually published anywhere so it probably doesn't really matter, but it might help to make it more clear in the codebase where macros are actually used if they are in a separate namespace.

mike-mcgann replied:
I actually did that at first but the code that uses them expect the macros to be in the same package and they are actually split across multiple packages with `lib` just being one of them. This might require some changes to the build.sbt. I think for this refactoring we should keep the functionality the same, and if we want to change this, it could be done in a separate PR.

stevedlawrence reviewed (commented):
:+1: , just some minor suggestions about new package names

This is also maybe semi related to [DAFFODIL-2406](https://issues.apache.org/jira/browse/DAFFODIL-2406). Though that's potentially a more invasive change and has some backwards compatibility issues so probably isn't worth doing now. But maybe something to keep in mind so that when the time comes that will be easier.

I imagine at some point we'll probably also want to take thorough look at our classes and make sure they are all really in the right back. I have a feeling things are sorted of scattered about, and the line between things like grammars and runtimes are maybe a bit blurry. Again, stuff for another day though.

mbeckerle commented on xsrc/test/scripts/refactor/lint-split-packages.scala:
```diff
@@ -0,0 +1,62 @@
+// import java.nio.file.Path
```
keeping this why? It's ok, just put comment at the top as to why.

mbeckerle reviewed (approved):
This looks good. I'm not surprised this turned into a multi-pass thing. 

We have to coordinate when to run this. 

Everyone will want to synchronize any outstanding changes they have anywhere with the commit just before when we apply this, to make updating for the new reorganized tree easier.

tuxji reviewed (changes requested):
I ran the refactor script and noticed the following things:

1. My scala was version 3.2.1 (I'd installed it with "sdk install scala" a long time ago but never used it to run a Scala program since I'd always run sbt and sbt uses the scala version in build.sbt).  The refactor script's scala commands printed errors and I had to git restore and git clean my checkout.  I removed scala 3.2.1 and installed scala 2.12.17.
2. The output from the bash scripts will be easier to understand if you run them with #!/bin/bash -x, so I've suggested these changes below.
3. The scripts changed the packages of the src/it tests.  I think their packages should have stayed the same since no one ever packages test classes in jars:

```diff
--- a/daffodil-cli/src/it/scala/org/apache/daffodil/CLI/Util.scala
+++ b/daffodil-cli/src/it/scala/org/apache/daffodil/CLI/Util.scala
@@ -15,7 +15,7 @@
 
-package org.apache.daffodil.CLI
+package org.apache.daffodil.cli.CLI
 
```

4. Since these api.DFDL classes live in runtime1, it makes it less likely that other runtimes/backends will implement these interfaces (IMHO, other runtimes/backends probably don't want to implement them anyway):

```
import org.apache.daffodil.runtime1.api.DFDL
import org.apache.daffodil.runtime1.api.DFDL.DataProcessor
import org.apache.daffodil.runtime1.api.DFDL.DaffodilUnparseErrorSAXException
import org.apache.daffodil.runtime1.api.DFDL.ParseResult
import org.apache.daffodil.runtime1.api.DFDL.UnparseResult
```

5. I ran sbt compile without any errors.
6. I ran sbt test without any errors.  It seems to me that the test classes still have their old packages, which I can understand since that minimizes the number of changes needed and no one ever packages test classes in jars.
7. I ran sbt IntegrationTest/test and it failed immediately:

```
sbt:daffodil> IntegrationTest/test
[info] Wrote /home/interran/apache/daffodil-mcgann/daffodil-cli/target/scala-2.12/daffodil-cli_2.12-3.5.0-SNAPSHOT.pom
[info] Main Scala API documentation to /home/interran/apache/daffodil-mcgann/daffodil-cli/target/scala-2.12/api...
[info] Main Scala API documentation successful.
[info] Main Scala API documentation to /home/interran/apache/daffodil-mcgann/daffodil-macro-lib/target/scala-2.12/api...
[info] Wrote /home/interran/apache/daffodil-mcgann/daffodil-macro-lib/target/scala-2.12/daffodil-macro-lib_2.12-3.5.0-SNAPSHOT.pom
[info] Main Scala API documentation successful.
[info] Wrote /home/interran/apache/daffodil-mcgann/daffodil-lib/target/scala-2.12/daffodil-lib_2.12-3.5.0-SNAPSHOT.pom
[info] Main Scala API documentation to /home/interran/apache/daffodil-mcgann/daffodil-lib/target/scala-2.12/api...
[error] /home/interran/apache/daffodil-mcgann/daffodil-lib/src/main/scala/org/apache/daffodil/lib/api/Validator.scala:22:1: Could not find any member to link for "org.apache.daffodil.validation.XercesValidator".
[error] /**
[error] ^
[error] one error found
[error] stack trace is suppressed; run last daffodil-lib / Compile / doc for the full output
[error] (daffodil-lib / Compile / doc) Scaladoc generation failed
[error] Total time: 23 s, completed Jan 19, 2023, 6:04:19 PM
sbt:daffodil>
```

8. I ran sbt ratCheck and it failed (the 11 licenses were the files in xsrc/test/refactor/scripts and the 5 binaries were files that weren't recognized as AL after the scripts changed them):

```
sbt:daffodil> ratCheck
[error] Unapproved licenses found: 11. See full report in /home/interran/apache/daffodil-mcgann/target/rat.txt
[error] Unapproved binaries found: 5. See full report in /home/interran/apache/daffodil-mcgann/target/rat.txt
[error] (ratCheck) org.musigma.sbt.rat.SbtRatCheck$UnapprovedLicenseException: Unapproved licenses
[error] Total time: 3 s, completed Jan 19, 2023, 6:07:21 PM
sbt:daffodil>
interran@GH3WPL13E:~/apache/daffodil-mcgann$ grep '  B ' /home/interran/apache/daffodil-mcgann/target/rat.txt | sort
  B     daffodil-japi/src/test/resources/META-INF/services/org.apache.daffodil.lib.api.ValidatorFactory
  B     daffodil-lib/src/main/resources/META-INF/services/org.apache.daffodil.lib.api.ValidatorFactory
  B     daffodil-lib/src/test/resources/META-INF/services/org.apache.daffodil.lib.api.ValidatorFactory
  B     daffodil-sapi/src/test/resources/META-INF/services/org.apache.daffodil.lib.api.ValidatorFactory
  B     daffodil-schematron/src/main/resources/META-INF/services/org.apache.daffodil.lib.api.ValidatorFactory
interran@GH3WPL13E:~/apache/daffodil-mcgann$
```

That's enough for this round.  I'll review again after some fixes are made.

mike-mcgann commented on xsrc/test/scripts/refactor/rename-dirs.sh:
```diff
@@ -0,0 +1,106 @@
+#!/bin/bash 
```
> 2. The output from the bash scripts will be easier to understand if you run them with #!/bin/bash -x, so I've suggested these changes below.

All the bash scripts should be updated with -ex and the main refactor.sh calls them using the shebang instead of explicitly through bash. 
 
> 3. The scripts changed the packages of the src/it tests.  I think their packages should have stayed the same since no one ever packages test classes in jars:

This got refactored in another commit so I think this has been overcome by events. 

>  7. I ran sbt IntegrationTest/test and it failed immediately:

The reference in the documentation has been updated along with the main class reference in the templates.

>  8. I ran sbt ratCheck and it failed (the 11 licenses were the files in xsrc/test/refactor/scripts and the 5 binaries were files that weren't recognized as AL after the scripts changed them):

I added licenses where they were missing. I do not know how to fix the services files into being recognized as text. I cannot seem to find where this is set in this codebase and looking at the rat repository didn't help either. Any help here would be appreciated.

mike-mcgann replied:
I worked with Steve on the ratCheck issue. After digging around a bit it appears that it was getting flagged as a binary because the filename contained ".lib". Those files have been placed in the RAT exclusion list for now as there doesn't seem to be an easy way to override that behavior.

mbeckerle commented:
In trying to run the refactor.sh script, I have found it is sensitive to the Java version and the Scala version that are executing at the command line. 

What versions are required? What versions are you using?

mike-mcgann commented:
> What versions are required? What versions are you using?

I'm running with Scala 2.13.10

mbeckerle commented:
Is there a way for a scala script to insist on a particular scala version, and error out if not the right one?

I just installed the default scala on my Ubuntu 20.04 Linux, and it is scala v2.11. So no wonder things broke.

mbeckerle commented:
Scripts work on scala 2.12.17, which is what I have installed now. 

I ran the scripts, and the resulting daffodil builds and tests without error. 

The CI did fail on two of the configurations, but I think that is spurious. 

I think we should start planning for when to merge this.  We need to give people a chance to settle any large changes that modify numerous files. (I have one sort of like that in flight that I want to think about. )

mike-mcgann commented:
`scripts/lint-split-packages.scala` should be working now. It should issue a bunch of errors when executed before the refactor and only a single warning after the refactor. I'm not sure what the best way is to incorporate this into the sbt build chain or in the CI if we wish to do that.

stevedlawrence commented:
> I'm not sure what the best way is to incorporate this into the sbt build chain or in the CI if we wish to do that.

One approach would be to create a new sbt task (e.g. sbt osgiCheck) and move the split package logic into the body of that task. I believe you could define the task an all logic in something like `project/OsgiCheck.scala` so it shouldn't require to much of a change. Another benefit of this approach is you can use other sbt tasks like `sources.value` to get a list of all scala and java source files--things like that might simplify the script a bit.

Another option, there might even be an existing sbt plugin that already does a similar check, seems like it would be useful to a number of projects to make sure they are OSGI compliant.

tuxji reviewed (changes requested):
I tested again.  I have suggestions and reservations about the refactor script.  All sbt commands I tried worked (compile, test, IntegrationTest/test, daffodil-cli/stage, ratCheck).  

I compared the contents of the Daffodil jar files in daffodil-cli/target/universal/stage/lib in my own checkout and the refactored checkout with `jar tf <jar> | sort` commands.  The refactored jars were identically named and most of their contents had identically named classes except for the insertion of a new package (a short name like cli, core, etc.) in the classes' folders.  However, there were some misses that may need to be corrected.

Run these commands with bash -x after refactoring and rebuilding Daffodil:

```
#!/bin/bash -x
jar tf daffodil-cli/target/universal/stage/lib/org.apache.daffodil.daffodil-cli-3.5.0-SNAPSHOT.jar | sort | grep -v META-INF | grep -v org/apache/daffodil/cli
jar tf daffodil-cli/target/universal/stage/lib/org.apache.daffodil.daffodil-core-3.5.0-SNAPSHOT.jar | sort | grep -v META-INF | grep -v org/apache/daffodil/core
jar tf daffodil-cli/target/universal/stage/lib/org.apache.daffodil.daffodil-io-3.5.0-SNAPSHOT.jar | sort | grep -v META-INF | grep -v org/apache/daffodil/io
jar tf daffodil-cli/target/universal/stage/lib/org.apache.daffodil.daffodil-japi-3.5.0-SNAPSHOT.jar | sort | grep -v META-INF | grep -v org/apache/daffodil/japi
jar tf daffodil-cli/target/universal/stage/lib/org.apache.daffodil.daffodil-lib-3.5.0-SNAPSHOT.jar | sort | grep -v META-INF | grep -v org/apache/daffodil/lib
jar tf daffodil-cli/target/universal/stage/lib/org.apache.daffodil.daffodil-runtime1-3.5.0-SNAPSHOT.jar | sort | grep -v META-INF | grep -v org/apache/daffodil/runtime1
jar tf daffodil-cli/target/universal/stage/lib/org.apache.daffodil.daffodil-runtime1-layers-3.5.0-SNAPSHOT.jar | sort | grep -v META-INF | grep -v org/apache/daffodil/layers/runtime1
jar tf daffodil-cli/target/universal/stage/lib/org.apache.daffodil.daffodil-runtime1-unparser-3.5.0-SNAPSHOT.jar | sort | grep -v META-INF | grep -v org/apache/daffodil/unparsers/runtime1
jar tf daffodil-cli/target/universal/stage/lib/org.apache.daffodil.daffodil-runtime2-3.5.0-SNAPSHOT.jar | sort | grep -v META-INF | grep -v org/apache/daffodil/runtime2
jar tf daffodil-cli/target/universal/stage/lib/org.apache.daffodil.daffodil-sapi-3.5.0-SNAPSHOT.jar | sort | grep -v META-INF | grep -v org/apache/daffodil/sapi
jar tf daffodil-cli/target/universal/stage/lib/org.apache.daffodil.daffodil-schematron-3.5.0-SNAPSHOT.jar | sort | grep -v META-INF | grep -v org/apache/daffodil/validation/schematron
jar tf daffodil-cli/target/universal/stage/lib/org.apache.daffodil.daffodil-tdml-lib-3.5.0-SNAPSHOT.jar | sort | grep -v META-INF | grep -v org/apache/daffodil/tdml
jar tf daffodil-cli/target/universal/stage/lib/org.apache.daffodil.daffodil-tdml-processor-3.5.0-SNAPSHOT.jar | sort | grep -v META-INF | grep -v org/apache/daffodil/tdml/processor
jar tf daffodil-cli/target/universal/stage/lib/org.apache.daffodil.daffodil-udf-3.5.0-SNAPSHOT.jar | sort | grep -v META-INF | grep -v org/apache/daffodil/udf
```

The daffodil-io jar contains these classes from other daffodil packages:

```
org/apache/daffodil/lib/exceptions/
org/apache/daffodil/lib/exceptions/AssertMacros$.class
org/apache/daffodil/lib/exceptions/AssertMacros.class
org/apache/daffodil/lib/exceptions/SDEMacros$.class
org/apache/daffodil/lib/exceptions/SDEMacros.class
org/apache/daffodil/lib/util/
org/apache/daffodil/lib/util/TimeTrackerMacros$.class
org/apache/daffodil/lib/util/TimeTrackerMacros.class
org/apache/daffodil/runtime1/processors/parsers/
org/apache/daffodil/runtime1/processors/parsers/PointOfUncertaintyMacros$$anon$1.class
org/apache/daffodil/runtime1/processors/parsers/PointOfUncertaintyMacros$$anon$2.class
org/apache/daffodil/runtime1/processors/parsers/PointOfUncertaintyMacros$.class
org/apache/daffodil/runtime1/processors/parsers/PointOfUncertaintyMacros.class
```

The daffodil-lib jar contains these classes from other daffodil packages:

```
org/apache/daffodil/io/
org/apache/daffodil/io/IOMacros$.class
org/apache/daffodil/io/IOMacros.class
org/apache/daffodil/runtime1/processors/parsers/
org/apache/daffodil/runtime1/processors/parsers/PointOfUncertaintyMacros$$anon$1.class
org/apache/daffodil/runtime1/processors/parsers/PointOfUncertaintyMacros$$anon$2.class
org/apache/daffodil/runtime1/processors/parsers/PointOfUncertaintyMacros$.class
org/apache/daffodil/runtime1/processors/parsers/PointOfUncertaintyMacros.class
```

The daffodil-runtime1 jar contains these classes from other daffodil packages:

```
org/apache/daffodil/io/
org/apache/daffodil/io/IOMacros$.class
org/apache/daffodil/io/IOMacros.class
org/apache/daffodil/lib/exceptions/
org/apache/daffodil/lib/exceptions/AssertMacros$.class
org/apache/daffodil/lib/exceptions/AssertMacros.class
org/apache/daffodil/lib/exceptions/SDEMacros$.class
org/apache/daffodil/lib/exceptions/SDEMacros.class
org/apache/daffodil/lib/util/
org/apache/daffodil/lib/util/TimeTrackerMacros$.class
org/apache/daffodil/lib/util/TimeTrackerMacros.class
```

The daffodil-tdml-lib jar contains these classes from other daffodil packages:

```
org/apache/daffodil/
org/apache/daffodil/Tak$.class
org/apache/daffodil/Tak.class
```

The daffodil-tdml-processor jar contains these classes from other daffodil packages:

```
org/apache/daffodil/tdml/
org/apache/daffodil/tdml/TDMLInfosetInputter.class
org/apache/daffodil/tdml/TDMLInfosetOutputter.class
```

tuxji commented:
P.S. I think it would be a good idea to run scalafmt after this refactoring with a very limited .scalafmt.conf which only sorts the import lines into the correct order.

stevedlawrence commented:
> I think it would be a good idea to run scalafmt after this refactoring with a very limited .scalafmt.conf which only sorts the import lines into the correct order.

Yeah. After this the rename is merged would be a good time to configure and run everything through scalafmt and get all the churn done at once.

Another thing we may want to do is add any rename and scalafmt commits to `.git-blame-ignore-revs` file. GitHub will use his in the blame view and local clones can be configured to use it with the blame.ignoreRevsFile. We would have to do it in a separate PR since merging PR's changes the hash, but we can do it once all the churn is done.

mike-mcgann commented:
> However, there were some misses that may need to be corrected.

I didn't include the TDML packages in the refactoring effort because I didn't think they would be loaded into an OSGI container but I went ahead and refactored these too and this is included in the recent updates. 

The macros seem to be a special case. They are in a separate package, span multiple packages, and get copied into the packages that use them. I'm not sure what we should be doing here with those. When I was talking to Mike B. he was interested in having these scripts do most of the heavy lifting and additional work to finalize could be done after. If you are okay with it, I think we should address the macros in a separate PR as that might require some research and additional discussion.

mike-mcgann commented on scripts/refactor/rename-dirs.sh:
```diff
@@ -0,0 +1,126 @@
+            mkdir -p $TEST_ROOT/$target_dir 
+            $MV $TEST_ROOT/* $TEST_ROOT/$target_dir || true
+            rm -rf $TEST_ROOT/$target_dir/$target_dir
+        fi 
```
Updated to rename via moving to a temporary directory first, creating the destination directory, and moving the contents to the final directory.

mike-mcgann commented on scripts/refactor/git-restore.sh:
```diff
@@ -0,0 +1,60 @@
+clean daffodil-macro-lib lib
+clean daffodil-runtime1 runtime1 
+# clean daffodil-runtime1-layers layers (SKIP)
+clean daffodil-runtime1-unparser unparsers
+
+rm -rf daffodil-io/src/{main,test}/scala/org/apache/daffodil/io/{layers,processors}
```
I wanted the git-restore script to be explicit so that I don't accidentally roll back the refactoring scripts themselves as I'm working on them.

tuxji replied:
Yes, I agree that is a good reason to be explicit and careful which files get rolled back!

tuxji commented:
> I think we should address the macros in a separate PR as that might require some research and additional discussion.

That's all right with me, and so is running scalafmt in separate PR(s) to re-sort the imports and/or reformat everything (the smaller the changes each PR makes at a time, the easier they will be to review).

mike-mcgann commented:
> One approach would be to create a new sbt task (e.g. sbt osgiCheck) and move the split package logic into the body of that task. I believe you could define the task an all logic in something like `project/OsgiCheck.scala` so it shouldn't require to much of a change. Another benefit of this approach is you can use other sbt tasks like `sources.value` to get a list of all scala and java source files--things like that might simplify the script a bit.

I moved the contents of the script into `project/OsgiCheck.scala` and added that as an sbt task too. I kept the filesystem traversal instead of using `sources.value` because it works already and that can always be updated when necessary.

mike-mcgann commented on project/Rat.scala:
```diff
@@ -38,6 +38,13 @@ object Rat {
+
     // test files that cannot include the Apache license without breaking tests
     file("daffodil-cli/src/test/resources/org/apache/daffodil/CLI/input/hextest.txt"),
     file("daffodil-cli/src/test/resources/org/apache/daffodil/CLI/input/input1.txt"),
```
Updated

mike-mcgann commented on scripts/refactor/fix-imports.scala:
```diff
@@ -0,0 +1,202 @@
+      symbols += s"${packageName}._" -> libName
+    }
+    symbols ++= Map(
+      "org.apache.daffodil.Main.ExitCode" -> "daffodil-cli",
+      "org.apache.daffodil.CLI.Util._" -> "daffodil-cli",
```
Updated, got pulled into the cliTest package.

tuxji reviewed (changes requested):
Reviewed again.  Noticed Rat.scala was included - it should be removed from the changeset and restored by git-restore.sh.  Want to suggest preserving daffodil-tdml-processor's package, see below.

stevedlawrence commented on project/OsgiCheck.scala:
```diff
@@ -0,0 +1,112 @@
+    nWarnings += 1
+  }
+
+  def checkPackage(fileName: String, libName: String, packageName: String): Unit = {
```
The way to handle this logic in a per subproject task could be to make two tasks. One is something like "osgiOwnedPackages" and it returns a `Seq[String]` of packages that each subproject owns. Then an "osgiCheck" task would just access that task from each subproject and make sure there are no overlaps.

stevedlawrence reviewed (commented):
Some suggestions on how to improve osgiCheck to be more sbt friendly.

mike-mcgann commented on project/OsgiCheck.scala:
```diff
@@ -0,0 +1,112 @@
+  val prog = "osgiCheck"
+
+  // Directories in the repository root that are libraries which should be checked
+  val libDirMatchers = Seq("""^.*(daffodil-\S+)$""".r)
```
I'm assuming that if the task is added to each subproject that the task could be run on a single library. The script needs to run over the entire source at once since it is looking for packages that are owned by *other* libraries. If you run osgiCheck against a single library it will always succeed because the owner map won't contain any packages from other libraries. So, does it make sense here to make it a task added at the subproject level?

mike-mcgann replied:
I was viewing a stale pageview of this PR and just saw your implementation below.

mike-mcgann commented on project/OsgiCheck.scala:
```diff
@@ -0,0 +1,112 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
```
I've replaced the OsgiCheck implementation with the one listed here

stevedlawrence commented on project/OsgiCheck.scala:
```diff
@@ -0,0 +1,96 @@
+        val optPackage = source.getLines.collectFirst {
+          case packageMatcher(packageName) => packageName
+        }
+        optPackage
```
The previous approach ignored with a warning, This just silently ignores, which feels wrong. Seems we either want to warn if optPackage is empty as before, or throw an exception if optPackage isn't defined and require they be fixed. We coukd also change the packageMatcher to allow commented out package lines which I think is the one case where optPackage is empty right now. I have no strong preference either way.

mike-mcgann replied:
I've made it a warning for now. We can always change it to an exception later if we feel that is better. 

I don't think the script should be looking at commented-out lines and it should be kept simple. There are many reasons that a package-ish looking statement could appear in comments and not be actual code.

mike-mcgann commented on scripts/refactor/rename-dirs.sh:
```diff
@@ -0,0 +1,152 @@
+
+rmdir daffodil-runtime1-unparser/$SRC_ROOT/processors
+
+move daffodil-tdml-processor/src/test/scala/org/apache/daffodil/tdml \
+     daffodil-tdml-processor/src/test/scala/org/apache/daffodil/processor/tdml
+sed -i 's/package org.apache.daffodil.tdml/package org.apache.daffodil.processor.tdml/' \
+    daffodil-tdml-processor/src/test/scala/org/apache/daffodil/processor/tdml/*
+move daffodil-tdml-processor/src/test/java/org/apache/daffodil/tdml \
+     daffodil-tdml-processor/src/test/java/org/apache/daffodil/processor/tdml
```
There was a similar discussion about package names towards the top of the thread. In a video call with Mike B., he said that we should just use what we have for now and package names can easily be refactored later. The key for these scripts is to do the heavy lifting that is too tedious with the IDE. Once the refactoring is close enough, the IDE can be used for any final polish. 

We can either keep these awkward names for now and rename them  later or we can do it now but I'll need consensus on what the new package names should be. I'm not familiar with the code and cannot provide any suggestions on what are good names here.

mike-mcgann commented on project/Rat.scala:
```diff
@@ -38,35 +38,42 @@ object Rat {
+    file("daffodil-cli/src/test/resources/org/apache/daffodil/cli/input/prefix.txt"),
+    file("daffodil-cli/src/test/resources/org/apache/daffodil/cli/input/test_DFDL-714.txt"),
+    file("daffodil-cli/src/test/resources/org/apache/daffodil/cli/input/uuid.txt"),
     file("daffodil-io/src/test/resources/iso8859.doc.dat"),
```
Reverted Rat.scala and added it to the git-restore script.

mike-mcgann commented on project/OsgiCheck.scala:
```diff
@@ -0,0 +1,112 @@
+    "daffodil-macro-lib", // multi-package
+  )
+  val ignorePackages = Set("org.apache.daffodil")
+  var ignoreDirs = Set("test")
```
I think we should just remove FieldFilder.scala from the source tree. The last update to it was five years ago. If we really want to use it later, we can grab it from git history. If we want to keep it around as documentation, we could change it to a markdown file.

mike-mcgann replied:
I went ahead and converted it to a markdown file. We can change it to something else if desired.

tuxji commented:
> Is ths plan for these scripts to be committed to the repo, or is this PR just for review and will be closed and new PR opened with the actual changes? 

Well, I would not assume that a refactoring of this scope is needed only once.  If we ever need to do another refactoring which is better done via scripts for reproducibility and review, these refactoring scripts written in both bash and scala have enough value and potential reuse that it's worth committing them to the repo and then removing them in the PR with the actual changes.  That way, the scripts will be preserved in at least one commit within git history for possible future reuse.

stevedlawrence commented:
> That way, the scripts will be preserved in at least one commit within git history for possible future reuse.

The scripts will always be preserved in this PR, though I understand that might be less obvious to find.

So seems reasonable to me. What are thoughts on doing the PR that makes the actual OSGI changes now, and once that review has the +1's we can then merge and revert this PR. That keeps the scripts in commit history, but allows us to make any changes if the actual OSGI changes reveals any issues that needs changes to the scripts?

tuxji commented:
> What are thoughts on doing the PR that makes the actual OSGI changes now, and once that review has the +1's we can then merge and revert this PR. That keeps the scripts in commit history, but allows us to make any changes if the actual OSGI changes reveals any issues that needs changes to the scripts?

Actually, this PR performs more changes than just adding the refactor scripts.  This PR also adds the osgiCheck command to build.sbt (with project/OsgiCheck.scala), changes FieldFinder/TestFieldFinder to markdown, and adds some new ignored files to Rat.scala.  We could simply delete the scripts directory in the still-to-come refactoring PR, otherwise all these additional changes need to be done either automatically by either the scripts or manually in the other PR instead of this PR.

stevedlawrence commented:
I would suggest that those changes (rat, osgiCheck, fieldfinder) be part of the OSGI pull request. That way we have a single commit for all the OSGI compliant changes and a single commit for the scripts. That way if anyone ever wants to look back at these scripts then they don't have to sift through a number of semi-related manual OSGI compliant changes.

mbeckerle commented:
> > Is ths plan for these scripts to be committed to the repo, or is this PR just for review and will be closed and new PR opened with the actual changes?
> 
> Well, I would not assume that a refactoring of this scope is needed only once. If we ever need to do another refactoring which is better done via scripts for reproducibility and review, these refactoring scripts written in both bash and scala have enough value and potential reuse that it's worth committing them to the repo and then removing them in the PR with the actual changes. That way, the scripts will be preserved in at least one commit within git history for possible future reuse.

I'd rather keep these scripts in the code tree someplace, and not delete them and have to search the git history to find them. Absent a tag, if they're only in the history they're gone AFAIC.  There are many more reasons why we may want to write scripts like this. I'd like to have a starting point of something that worked in the past without having to hunt for it.

stevedlawrence commented:
> I'd rather keep these scripts in the code tree someplace, and not delete them and have to search the git history to find them.

Fair enough. In that case, can we rename the scripts directory to `scripts/osgi-refactor` to make it clear what the purpose is and differentiate if we ever need to build some other kind of refactoring scripts?

Also, should we include the actual OSGI changes in this PR so there's single commit related to osgi refactoring?

stevedlawrence commented:
I'm also fine if we want to keep OSGI refactor scripts and OSGI compliance changes in separate commits in one PR, but I think they should be reviewed at the same time since they are really one in the same.

mike-mcgann commented:
> In that case, can we rename the scripts directory to scripts/osgi-refactor

Done

> I'm also fine if we want to keep OSGI refactor scripts and OSGI compliance changes in separate commits in one PR, but I think they should be reviewed at the same time since they are really one in the same

If we are going to keep the scripts around, I think the easiest would be to bundle them up into a single PR.

stevedlawrence commented:
Yeah, I was suggesting a single PR containing two commits, one commit for the scripts and one commit for the actual changes to make things OSGI compliant (made either manually or via scripts). Makes the commit history more clear about what is what. Also, if we use [.git-blame-ignore-revs](https://github.blog/changelog/2022-03-24-ignore-commits-in-the-blame-view-beta/) (which we probably should), then we only ignore all the move/rename stuff and not adding actual new files.

tuxji commented:
Splitting the PR into two commits with all the move/rename changes in their own separate commit and [ignoring that commit in the blame view](https://docs.github.com/en/repositories/working-with-files/using-files/viewing-a-file#ignore-commits-in-the-blame-view) sounds good to me.   For the purpose of the git blame view, I think it makes sense to include the FieldFinder, TestFieldFinder, and Rat.scala changes in the move/rename commit and put the build.sbt & OsgiCheck.scala changes in the scripts commit.

However, we need to do the move/rename changes in the first commit because we need to put the move/rename commit's hash in the .git-blame-ignore-revs file.  @mike-mcgann, can you run the necessary git fetch/git rebase -i/git push --force-with-lease commands to rearrange your PR into that order (first commit doing the move/rename changes, second commit saving the scripts and putting the first commit's hash in .git-blame-ignore-revs)?  Then we can review the PR for final approval before merging it.

mike-mcgann commented:
> first commit doing the move/rename changes, second commit saving the scripts

I would prefer the opposite order if that is okay--first commit is the scripts, second commit is the refactoring. If we decide to make further changes at this point, I can then discard HEAD, update HEAD~1, commit, run, commit. And easier for now too. I can rebase, verify, commit, run, commit. Otherwise I have to ignore the scripts, commit, add the scripts, run, commit.

tuxji commented:
> I would prefer the opposite order if that is okay--first commit is the scripts, second commit is the refactoring. 

Yes, your reasons make sense.  Like Steve said, we can punt .git-blame-ignore-revs to a future PR anyway.

mike-mcgann commented:
I rebased, commited the scripts, ran the refactor, committed that and fixed a conflict. Ready for final review.

tuxji commented:
I took a quick look.  I ran these commands in my terminal:

```shell
interran@GH3WPL13E:~$ cd apache/daffodil-mcgann/
interran@GH3WPL13E:~/apache/daffodil-mcgann$ git pull
Fetching origin
remote: Enumerating objects: 1901, done.
remote: Counting objects: 100% (1860/1860), done.
remote: Compressing objects: 100% (1101/1101), done.
remote: Total 1901 (delta 474), reused 1750 (delta 440), pack-reused 41
Receiving objects: 100% (1901/1901), 2.38 MiB | 3.92 MiB/s, done.
Resolving deltas: 100% (474/474), completed with 166 local objects.
From github.com:mike-mcgann/daffodil
 + 10e5daf65...9937fe68a daffodil-2683-unsplit-packages   -> origin/daffodil-2683-unsplit-packages  (forced update)
 * [new branch]          daffodil-2133-scalafmt           -> origin/daffodil-2133-scalafmt
 * [new branch]          daffodil-2676.leading_whitespace -> origin/daffodil-2676.leading_whitespace
   2b8e52b82..10c520aef  main                             -> origin/main
fatal: Not possible to fast-forward, aborting.
interran@GH3WPL13E:~/apache/daffodil-mcgann$ git switch daffodil-2683-unsplit-packages
Already on 'daffodil-2683-unsplit-packages'
Your branch and 'origin/daffodil-2683-unsplit-packages' have diverged,
and have 28 and 25 different commits each, respectively.
  (use "git pull" to merge the remote branch into yours)
interran@GH3WPL13E:~/apache/daffodil-mcgann$ git reset --hard origin/daffodil-2683-unsplit-packages
HEAD is now at 9937fe68a Refactor for OSGI compliance
interran@GH3WPL13E:~/apache/daffodil-mcgann$ git log
commit 9937fe68a44b4ba731e9cdfd58e1c1e93532acc7 (HEAD -> daffodil-2683-unsplit-packages, origin/daffodil-2683-unsplit-packages)
Author: Mike McGann <mmcgann@owlcyberdefense.com>
Date:   Fri Feb 3 13:10:25 2023 -0500

    Refactor for OSGI compliance

    Daffodil shares package namespaces across different libraries such that
    library A can have a class named foo.bar.One and library B can have a
    class named foo.bar.Two. This can cause issues with class loaders that
    prohibit different jars from owning the same package (i.e., in an
    OSGI container). This factoring ensures that each package is owned by
    a single jar.

    DAFFODIL-2683

commit 5f778d7084de722a5fcedfae5bfddfa40cdc572f
Author: Mike McGann <mmcgann@owlcyberdefense.com>
Date:   Tue Jan 10 15:37:27 2023 -0500

    Add refactor scripts for OSGI compliance

    Daffodil shares package namespaces across different libraries such that
    library A can have a class named foo.bar.One and library B can have a
    class named foo.bar.Two. This can cause issues with class loaders that
    prohibit different jars from owning the same package (i.e., in an
    OSGI container). These scripts refactor the code to ensure that each
    package is only owned by a single jar. An sbt task, osgiCheck, has
    been added to enforce compilance in the future.

    DAFFODIL-2683

commit d62be6482c8594e8940832005401408b64c169dd
Author: Steve Lawrence <slawrence@apache.org>
Date:   Thu Feb 2 12:27:52 2023 -0500

    Reduce the maximum length of xs:hexBinary to Int.MaxValue / 2
...
interran@GH3WPL13E:~/apache/daffodil-mcgann$ git diff HEAD~2..HEAD~
...
```

The first commit's diff looks good except for one thing.  Rat.scala is changed in both this commit and the next commit.  It probably should be changed only in the second commit, but if no one else objects, it's not a big deal.

```shell
interran@GH3WPL13E:~/apache/daffodil-mcgann$ git diff HEAD~2..HEAD~ project/Rat.scala
...
interran@GH3WPL13E:~/apache/daffodil-mcgann$ git diff HEAD~..HEAD project/Rat.scala
...
```

You also may need to configure your IDE so that it always appends a newline as the final character in files.  My git diff complained about there being no newline at the end of the following files:

```
daffodil-runtime1/src/main/scala/org/apache/daffodil/reflection/FieldFinder.scala.md
daffodil-runtime1/src/test/scala/org/apache/daffodil/reflection/TestFieldFinder.scala.md
project/OsgiCheck.scala
scripts/osgi-refactor/.gitignore
scripts/osgi-refactor/fix-imports.scala
```

If Mike asks you to rework something in the first commit, please add newlines to these files.

The second commit's diffs look like what I saw before, although I noticed this time that some methods had a local `import NodeInfo._` statement in their bodies which was turned into an `import org.apache.daffodil.runtime1.dpath.NodeInfo._` even though there was already an `import org.apache.daffodil.runtime1.dpath.NodeInfo` at the beginning of these files.  These methods could have kept their original `import NodeInfo._` local statements without needing to qualify the import.  I also saw one `import InfosetEventKind._` get turned into `import org.apache.daffodil.runtime1.infoset.InfosetEventKind._`.

Finally, the second commit also changes one line in one of the refactoring scripts, see `git diff HEAD~..HEAD scripts/osgi-refactor/refactor.sh`.  That change should've been in the first commit.  Note that I make it a habit to look at my own branch's git diff output and write a long commit message describing why I changed *each* file, which helps me catch unintended changes or rethink some changes to make them better.

OK, it's Mike's turn now :).

tuxji commented on daffodil-core/src/main/scala/org/apache/daffodil/core/dpath/Conversions.scala:
```diff
@@ -42,7 +44,7 @@ object Conversion {
    * output type. TT is the target type desired.
    */
   def conversionOps(st: NodeInfo.Kind, tt: NodeInfo.Kind, context: Expression): List[RecipeOp] = {
-    import NodeInfo._
+    import org.apache.daffodil.runtime1.dpath.NodeInfo._
```
Here's one of the local `import NodeInfo._` statements.

mbeckerle reviewed (approved):
+1

The title of this PR should be changed as it is not just about creating scripts. 

I pulled your branch down, because review using github for this is infeasible.

mike-mcgann commented:
> The first commit's diff looks good except for one thing. Rat.scala is changed in both this commit and the next commit. 

I went back and fixed that so that the script adds them. 

> You also may need to configure your IDE so that it always appends a newline as the final character in files. My git diff complained about there being no newline at the end of the following files:

I fixed the settings in both editors. Hopefully scalafmt will take care of that in the future. 

> The second commit's diffs look like what I saw before, although I noticed this time that some methods had a local `import NodeInfo._` statement in their bodies which was turned into an `import org.apache.daffodil.runtime1.dpath.NodeInfo._` even though there was already an `import org.apache.daffodil.runtime1.dpath.NodeInfo` at the beginning of these files. These methods could have kept their original `import NodeInfo._` local statements without needing to qualify the import. I also saw one `import InfosetEventKind._` get turned into `import org.apache.daffodil.runtime1.infoset.InfosetEventKind._`.

I removed those from the script.

stevedlawrence commented:
The first commit also includes the FieldFinder changes and the OSGI check pluging. Should either/both of those be in the second commit? I could see arguments either way so don't feel that strongly about it either way but thought it was worth consideration.

tuxji reviewed (approved):
> The first commit also includes the FieldFinder changes and the OSGI check plugin. Should either/both of those be in the second commit? I could see arguments either way so don't feel that strongly about it either way but thought it was worth consideration.

I have just finished reviewing a checkout of the PR branch so I would say no, let's not have another go-around.  I even filtered out the packages/imports changes from my diff to look for any unexpected changes.  I saw the private[tdml] removed from a RunnerFactory method to fix a compilation error and all the other changes were whitespace changes (final newline, or blank lines with whitespace changes) or occurrences of org.apache.daffodil.<name> outside of packages/imports.  +1 to merging this PR (Steve, I'll let you do the honors since you're the last reviewer).

stevedlawrence reviewed (approved):
+1 :+1:
